gcpkms/keysource.go

```diff
@@ -245,12 +248,21 @@ func (key *MasterKey) newKMSClient() (*kms.KeyManagementClient, error) {
 	default:
 		credentials, err := getGoogleCredentials()
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("credentials: failed to obtain credentials from %q: %w", SopsGoogleCredentialsEnv, err)
 		}
 		if credentials != nil {
 			opts = append(opts, option.WithCredentialsJSON(credentials))
+			break
 		}
+
+		if atCredentials := getGoogleOAuthTokenFromEnv(); atCredentials != nil {
+			opts = append(opts, option.WithTokenSource(atCredentials))
+			break
+		}
+
+		return nil, fmt.Errorf("credentials: none found: missing %q or %q environment variable", SopsGoogleCredentialsEnv, SopsGoogleCredentialsOAuthTokenEnv)
```

Comment:
Hmm, the old behavior was to not fail if no credentials have been provided in the `default` branch. (Whether that makes sense - no idea. Maybe the GCP SDK does something else in that case to get hold of credentials?) If we want to restore this:
```suggestion
		}
```

Reply:
> Maybe the GCP SDK does something else in that case to get hold of credentials?

Yes, it does. All the google libraries default to [Application Default Credentials](https://cloud.google.com/docs/authentication/application-default-credentials), which is essentially searching for credentials in the environment:

* The `GOOGLE_APPLICATION_CREDENTIALS` standard environment variable (btw none of the other env vars here are this one)
* A file created in a well-known location by the command `gcloud auth application-default login`
* The GCE/GKE metadata server

However, `getGoogleCredentials()` (which should probably be called `getGoogleCredentialsFromEnv()`), only returns an error if the non-standard `GOOGLE_CREDENTIALS` environment variable is set, and this variable is only set when the user is declaring a clear intention of using it, so erroring out the entire operation makes sense to me. If the code ends up using an unintended authentication method that is coincidentally available, that's pretty bad in my opinion. People need to organize their configuration well, this new version of the code should help them out on that. Disruptive? Yes, probably, but for a greater good in my opinion.

Reply:
> Disruptive? Yes, probably, but for a greater good in my opinion.

It is still a breaking change, I'd prefer not to introduce these in a minor release.

Reply:
Oh nevermind, for some reason I wrapped my head around the error returned from `getGoogleCredentials()`, and not this one on the bottom. Yes, this comments makes a lot of sense!